source/common/http/http1/codec_impl.cc

```diff
@@ -892,6 +902,10 @@ void ServerConnectionImpl::onMessageComplete() {
   ASSERT(!handling_upgrade_);
   if (active_request_.has_value()) {
     auto& active_request = active_request_.value();
+
+    if (active_request.request_decoder_) {
+      active_request.response_encoder_.readDisable(true);
```

Comment:
@alyssawilk the reason to 'readDisable' here is to avoid the client that emit pipelined requests, in this case, the envoy may loop handling the http requests only from this client and starve other clients?

Reply:
It's not about looping on requests from a single client, Envoy's HTTP/1.1 codec (as with most HTTP/1.1 code) doesn't actually handle pipelined requests simultaneously.  It always processes the first request fully before starting on a second, since with HTTP/1.1 any error with the first request (sending a 400 before a POST body is read, server time out etc.) may result in the connection getting closed.

Reply:
@alyssawilk thanks for your reply :) 

I wonder if we can make sure the client won't make pipelined requests then we can remove this line of code without any negative impact?

On the other hand, what is the negative impact if the client make pipelined requests but we remove this line of code?

Reply:
Unfortunately as Envoy is often deployed as an edge proxy, with random internet clients, there's no way we can make that guarantee.  Reading a subsequent request without having fully processed the first would likely cause crashes.
Can you give some insight into why you want to remove this line?

Reply:
It seems has some negative impact for performance, maybe too frequent event operation? One event add/remove to libevent for on http request?

And I wonder for sidecar of service mesh? Could we remove this "read disable" operation to get better performance? Even disable watermark, after all, the scene is more under control :)

Reply:
Huh, so I'm surprised this particular one shows up for perf - can you share flame graphs or anything?
if it is a legitimate problem and you're confident it never happens for your deployments, you could add a configuration option (default off) to treat pipelined requests as fatal and in lieu of read disabling leave it open and close the connection if there's unexpected data.  It'd be a high risk change though so it'd help to have perf data to justify it.

Reply:
There's an internal bug about "5% to 10% of processing time for small HTTP/1.1 requests in ConnectionImpl::readDisable" that I wrote long ago, but haven't been able to justify actions on due to lack of loadtests that show the issue. I think there's room for optimization here by avoiding the file event mask changes in cases where the codec is not interested in reading. The other challenge in making that optimization is in the use of the CLOSE event signal which only works in Linux. It would be helpful to track this performance issue in some way.

Reply:
I use fortio to do test with 100 connections and no qps limit, ref:

`
fortio load -qps -1 -c 100 -t 10s --timeout 120s  http://127.0.0.1:10001
`

And I deploy three envoys,  two as proxys which do simple forwarding and one as application which do direct response:

`
fortio -> client side envoy proxy -> server side envoy proxy -> envoy as application
`

And for envoy proxy with read disable, the result qps is 11000 on average. But for envoy proxy without read disalbe, the result qps is 15000 on average.